panic: set user priority called on leaf txn

[(roachtest).sqlsmith/setup=rand-tables/setting=no-ddl failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1799071&tab=buildLog) on [master@72c4a1bd411f2f82bf9aaa22883821a946614148](https://github.com/cockroachdb/cockroach/commits/72c4a1bd411f2f82bf9aaa22883821a946614148):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200310-1799071/sqlsmith/setup=rand-tables/setting=no-ddl/run_1
	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:801: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1799071-1583824469-11-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: 5839
		4: 5499
		2: 6148
		3: dead
		Error:  3: dead

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-ddl](https://teamcity.cockroachdb.com/viewLog.html?buildId=1799071&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-ddl)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-ddl.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

Node 3 died with this stack trace:
```
panic: SetUserPriority() called on leaf txn

goroutine 12851 [running]:
panic(0x3f36ce0, 0xc004f0a260)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc004fa3e00 sp=0xc004fa3d70 pc=0x799232
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).SetUserPriority(0xc003ea30e0, 0x3ff0000000000000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:204 +0x2ad fp=0xc004fa3ec0 sp=0xc004fa3e00 pc=0x13286bd
github.com/cockroachdb/cockroach/pkg/sql.(*txnState).setPriority(0xc004dcd090, 0x3ff0000000000000, 0xc0044f2c40, 0xc003e04280)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/txn_state.go:298 +0x4f fp=0xc004fa3f00 sp=0xc004fa3ec0 pc=0x2d05cff
github.com/cockroachdb/cockroach/pkg/sql.(*txnState).resetForNewSQLTxn(0xc004dcd090, 0x4b45020, 0xc005200c00, 0x1, 0xe311661, 0xed5f938e5, 0x0, 0x0, 0x3ff0000000000000, 0x2, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/txn_state.go:211 +0x416 fp=0xc004fa3fe0 sp=0xc004fa3f00 pc=0x2d050a6
github.com/cockroachdb/cockroach/pkg/sql.(*Server).newConnExecutorWithTxn(0xc000b59180, 0x4b45020, 0xc005200c00, 0xc00374bd40, 0xc0044f2b80, 0xc0042275c0, 0x4bb7380, 0xc005200cf0, 0xc003e04140, 0xc0005f4ee0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:712 +0x3e5 fp=0xc004fa4150 sp=0xc004fa3fe0 pc=0x2be8085
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).initConnEx(0xc004227560, 0x4b45020, 0xc005200c00, 0xc003ea30e0, 0xc00374bd40, 0xc00374bd40, 0x0, 0xc00003ea80, 0x4abdca0, 0x74a2738, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:137 +0x230 fp=0xc004fa42a0 sp=0xc004fa4150 pc=0x2c7c650
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal(0xc004227560, 0x4b45020, 0xc005200c00, 0x42af484, 0x10, 0xc003ea30e0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:424 +0x687 fp=0xc004fa4848 sp=0xc004fa42a0 pc=0x2c7e257
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternal(0xc004227560, 0x4b45020, 0xc005200ab0, 0x42af484, 0x10, 0xc003ea30e0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:224 +0xef fp=0xc004fa4980 sp=0xc004fa4848 pc=0x2c7d00f
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryEx(0xc004227560, 0x4b45020, 0xc005200ab0, 0x42af484, 0x10, 0xc003ea30e0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:199 +0xec fp=0xc004fa4a58 sp=0xc004fa4980 pc=0x2c7cd1c
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx(0xc004227560, 0x4b45020, 0xc005200ab0, 0x42af484, 0x10, 0xc003ea30e0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:254 +0xec fp=0xc004fa4b18 sp=0xc004fa4a58 pc=0x2c7d57c
github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRow(0xc004227560, 0x4b45020, 0xc005200ab0, 0x42af484, 0x10, 0xc003ea30e0, 0xc0006a20f0, 0x4e, 0xc004f0a170, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/internal.go:238 +0x195 fp=0xc004fa4c68 sp=0xc004fa4b18 pc=0x2c7d2f5
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*lazyInternalExecutor).QueryRow(0xc005200b40, 0x4b45020, 0xc005200ab0, 0x42af484, 0x10, 0xc003ea30e0, 0xc0006a20f0, 0x4e, 0xc004f0a170, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:610 +0xf9 fp=0xc004fa4d08 sp=0xc004fa4c68 pc=0x2b3c799
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.queryOidWithJoin(0xc0046668c0, 0x6daeea0, 0x4bbf900, 0xc004154400, 0x0, 0x0, 0x0, 0x0, 0x424b700, 0xc000640640, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3189 +0x408 fp=0xc004fa4ee0 sp=0xc004fa4d08 pc=0x1364ea8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.queryOid(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3212
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.PerformCast(0xc0046668c0, 0x4bbf900, 0xc004154400, 0x6daeea0, 0xc004154400, 0x0, 0xc005200b70, 0x4b5d1e0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3723 +0x4a35 fp=0xc004fa5430 sp=0xc004fa4ee0 pc=0x1369f15
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*CastExpr).Eval(0xc005200b70, 0xc0046668c0, 0xc005200b70, 0x4b80201, 0xc005200b70, 0x4b5da01)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3227 +0xd7 fp=0xc004fa5490 sp=0xc004fa5430 pc=0x1365417
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ConstantEvalVisitor).VisitPost(0xc005200bd0, 0x4b5d1e0, 0xc005200b70, 0x4b5d1e0, 0xc005200b70)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/constant_eval.go:50 +0x98 fp=0xc004fa54f8 sp=0xc004fa5490 pc=0x1345268
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.WalkExpr(0x4afe9a0, 0xc005200bd0, 0x4b5d1e0, 0xc005200b70, 0xc005200b70, 0xc005200b70, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/walk.go:684 +0x233 fp=0xc004fa5558 sp=0xc004fa54f8 pc=0x13d4ad3
github.com/cockroachdb/cockroach/pkg/sql/execinfra.processExpression(0x0, 0x0, 0xc005124900, 0x26, 0x0, 0x0, 0xc0046668c0, 0xc0043e00d0, 0xc004e08710, 0x36, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/expr.go:87 +0x2cb fp=0xc004fa5600 sp=0xc004fa5558 pc=0x20aeb0b
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ExprHelper).Init(0xc004e08700, 0x0, 0x0, 0xc005124900, 0x26, 0x0, 0x0, 0xc004663500, 0x11, 0x11, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/expr.go:161 +0x293 fp=0xc004fa57c8 sp=0xc004fa5600 pc=0x20af223
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcOutputHelper).Init(0xc004e08018, 0xc003cd8260, 0xc004663500, 0x11, 0x11, 0xc0046668c0, 0x4b249e0, 0xc004e4d880, 0xc004663501, 0x74a2738)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:145 +0x816 fp=0xc004fa5980 sp=0xc004fa57c8 pc=0x20b2d66
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).InitWithEvalCtx(0xc004e08000, 0x4b69160, 0xc004e08000, 0xc003cd8260, 0xc004663500, 0x11, 0x11, 0xc00374bc20, 0xc0046668c0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:800 +0x135 fp=0xc004fa59e0 sp=0xc004fa5980 pc=0x20b5555
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Init(0xc004e08000, 0x4b69160, 0xc004e08000, 0xc003cd8260, 0xc004663500, 0x11, 0x11, 0xc00374bc20, 0x0, 0x4b249e0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:776 +0x243 fp=0xc004fa5aa8 sp=0xc004fa59e0 pc=0x20b53a3
github.com/cockroachdb/cockroach/pkg/sql/rowexec.newTableReader(0xc00374bc20, 0x0, 0xc00395e900, 0xc003cd8260, 0x4b249e0, 0xc004e4d880, 0x0, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:93 +0x22a fp=0xc004fa5c00 sp=0xc004fa5aa8 pc=0x2105dea
github.com/cockroachdb/cockroach/pkg/sql/rowexec.NewProcessor(0x4b44f60, 0xc0044f2a40, 0xc00374bc20, 0x0, 0xc003cd8198, 0xc003cd8260, 0x0, 0x0, 0x0, 0xc0040a0fa8, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/processors.go:138 +0x2cdc fp=0xc004fa6e90 sp=0xc004fa5c00 pc=0x20ec19c
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).makeProcessor(0xc0032b6e30, 0x4b44f60, 0xc0044f2a40, 0xc003cd8180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:228 +0x1dd fp=0xc004fa6fc8 sp=0xc004fa6e90 pc=0x2b32bed
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).setupProcessors(0xc0032b6e30, 0x4b44f60, 0xc0044f2a40, 0xc0000135a0, 0xc004154380, 0x1, 0x1, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:81 +0x1af fp=0xc004fa70c8 sp=0xc004fa6fc8 pc=0x2b3267f
github.com/cockroachdb/cockroach/pkg/sql/rowflow.(*rowBasedFlow).Setup(0xc0032b6e30, 0x4b44f60, 0xc0044f2a40, 0xc0000135a0, 0x4b45000, 0xc000ac8630, 0xc0000e5200, 0x264842bb367276bc, 0xa3df17f51cbdca8b)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowflow/row_based_flow.go:64 +0x13f fp=0xc004fa7128 sp=0xc004fa70c8 pc=0x2b3244f
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).setupFlow(0xc0000e5200, 0x4b45020, 0xc000ac8630, 0x0, 0x0, 0xc0000e5398, 0xc000013590, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:350 +0x5b8 fp=0xc004fa75d0 sp=0xc004fa7128 pc=0x2b39968
github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow(0xc0000e5200, 0x4b45020, 0xc005200960, 0xc000013590, 0xc0000e5200, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:518 +0x20b fp=0xc004fa76c8 sp=0xc004fa75d0 pc=0x2b3babb
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1(0x4b45020, 0xc005200960, 0x40fe940, 0xc000013590, 0x40fe940, 0xc000013590, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:630 +0x86 fp=0xc004fa7710 sp=0xc004fa76c8 pc=0x20a4f66
github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x4b45020, 0xc005200960, 0x40fe940, 0xc000013590, 0xc0041542e0, 0xc004154300, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0x9ef fp=0xc004fa79b0 sp=0xc004fa7710 pc=0x15abdef
github.com/cockroachdb/cockroach/pkg/rpc.NewServerWithInterceptor.func2(0x4b45020, 0xc005200960, 0x40fe940, 0xc000013590, 0xc0041542e0, 0xc004154300, 0xc0038bca78, 0x95ec78, 0x4042da0, 0xc005200960)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:223 +0xb6 fp=0xc004fa7a18 sp=0xc004fa79b0 pc=0x15bb576
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler(0x41c6ea0, 0xc0000e5200, 0x4b45020, 0xc005200960, 0xc004227440, 0xc000309cc0, 0x4b45020, 0xc005200960, 0xc004dca800, 0x784)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:632 +0x14b fp=0xc004fa7a88 sp=0xc004fa7a18 pc=0x2021c8b
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc00041fc80, 0x4b9ec80, 0xc00354e300, 0xc003b36b00, 0xc000cb2c60, 0x6d88de0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:999 +0x460 fp=0xc004fa7e18 sp=0xc004fa7a88 pc=0xe65cf0
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).handleStream(0xc00041fc80, 0x4b9ec80, 0xc00354e300, 0xc003b36b00, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1279 +0xd97 fp=0xc004fa7f48 sp=0xc004fa7e18 pc=0xe69cf7
github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00030d630, 0xc00041fc80, 0x4b9ec80, 0xc00354e300, 0xc003b36b00)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:717 +0xbb fp=0xc004fa7fb8 sp=0xc004fa7f48 pc=0xe7666b
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc004fa7fc0 sp=0xc004fa7fb8 pc=0x7c98a1
created by github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:715 +0xa1
```

I think @andreimatei has some context on this?

--- a/pkg/kv/kvclient/kvcoord/txn_coord_sender.go
+++ b/pkg/kv/kvclient/kvcoord/txn_coord_sender.go
@@ -1122,10 +1122,6 @@ func (tc *TxnCoordSender) PrepareRetryableError(ctx context.Context, msg string)
 
 // Step is part of the TxnSender interface.
 func (tc *TxnCoordSender) Step(ctx context.Context) error {
-	if tc.typ != kv.RootTxn {
-		return errors.WithContextTags(
-			errors.AssertionFailedf("cannot call Step() in leaf txn"), ctx)
-	}
 	tc.mu.Lock()
 	defer tc.mu.Unlock()
 	return tc.interceptorAlloc.txnSeqNumAllocator.stepLocked(ctx)
@@ -1135,10 +1131,6 @@ func (tc *TxnCoordSender) Step(ctx context.Context) error {
 func (tc *TxnCoordSender) ConfigureStepping(
 	ctx context.Context, mode kv.SteppingMode,
 ) (prevMode kv.SteppingMode) {
-	if tc.typ != kv.RootTxn {
-		panic(errors.WithContextTags(
-			errors.AssertionFailedf("cannot call ConfigureStepping() in leaf txn"), ctx))
-	}
 	tc.mu.Lock()
 	defer tc.mu.Unlock()
 	return tc.interceptorAlloc.txnSeqNumAllocator.configureSteppingLocked(mode)

--- a/pkg/kv/txn.go
+++ b/pkg/kv/txn.go
@@ -1199,10 +1199,6 @@ func (txn *Txn) ManualRestart(ctx context.Context, ts hlc.Timestamp) {
 // operation (usually, but not exclusively, by a high-priority txn with
 // conflicting writes).
 func (txn *Txn) IsSerializablePushAndRefreshNotPossible() bool {
-	if txn.typ != RootTxn {
-		panic(
-			errors.AssertionFailedf("IsSerializablePushAndRefreshNotPossible() called on leaf txn"))
-	}
 	return txn.mu.sender.IsSerializablePushAndRefreshNotPossible()
 }
 
@@ -1246,10 +1242,6 @@ func (txn *Txn) Active() bool {
 // In step-wise execution, reads operate at a snapshot established at
 // the last step, instead of the latest write if not yet enabled.
 func (txn *Txn) Step(ctx context.Context) error {
-	if txn.typ != RootTxn {
-		return errors.WithContextTags(
-			errors.AssertionFailedf("txn.Step() only allowed in RootTxn"), ctx)
-	}
 	txn.mu.Lock()
 	defer txn.mu.Unlock()
 	return txn.mu.sender.Step(ctx)
@@ -1258,10 +1250,6 @@ func (txn *Txn) Step(ctx context.Context) error {
 // ConfigureStepping configures step-wise execution in the
 // transaction.
 func (txn *Txn) ConfigureStepping(ctx context.Context, mode SteppingMode) (prevMode SteppingMode) {
-	if txn.typ != RootTxn {
-		panic(errors.WithContextTags(
-			errors.AssertionFailedf("txn.ConfigureStepping() only allowed in RootTxn"), ctx))
-	}
 	txn.mu.Lock()
 	defer txn.mu.Unlock()
 	return txn.mu.sender.ConfigureStepping(ctx, mode)

--- a/pkg/sql/conn_executor.go
+++ b/pkg/sql/conn_executor.go
@@ -695,6 +695,11 @@ func (s *Server) newConnExecutorWithTxn(
 	appStats *appStats,
 ) *connExecutor {
 	ex := s.newConnExecutor(ctx, sd, sdDefaults, stmtBuf, clientComm, memMetrics, srvMetrics, appStats)
+	if txn.Type() == kv.LeafTxn {
+		// If the txn is a leaf txn it is not allowed to perform mutations. For
+		// sanity, set read only on the session.
+		ex.dataMutator.SetReadOnly(true)
+	}
 
 	// The new transaction stuff below requires active monitors and traces, so
 	// we need to activate the executor now.
@@ -712,7 +717,7 @@ func (s *Server) newConnExecutorWithTxn(
 		explicitTxn,
 		txn.ReadTimestamp().GoTime(),
 		nil, /* historicalTimestamp */
-		txn.UserPriority(),
+		roachpb.UnspecifiedUserPriority,
 		tree.ReadWrite,
 		txn,
 		ex.transitionCtx)

--- a/pkg/sql/internal_test.go
+++ b/pkg/sql/internal_test.go
@@ -17,7 +17,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/server"
 	"github.com/cockroachdb/cockroach/pkg/sql"
@@ -458,6 +460,26 @@ func testInternalExecutorAppNameInitialization(
 	} else if appName := string(*rows[0][0].(*tree.DString)); appName != expectedAppNameInStats {
 		t.Fatalf("unexpected app name: expected %q, got %q", expectedAppNameInStats, appName)
 	}
+}
+
+func TestInternalExecutorInLeafTxnDoesNotPanic(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	s, _, kvDB := serverutils.StartServer(t, base.TestServerArgs{})
+	defer s.Stopper().Stop(ctx)
+
+	rootTxn := kvDB.NewTxn(ctx, "root-txn")
+
+	ltis := rootTxn.GetLeafTxnInputState(ctx)
+	leafTxn := kv.NewLeafTxn(ctx, kvDB, roachpb.NodeID(1), &ltis)
+
+	ie := s.InternalExecutor().(*sql.InternalExecutor)
+	_, err := ie.QueryEx(
+		ctx, "leaf-query", leafTxn, sessiondata.InternalExecutorOverride{User: security.RootUser}, "SELECT 1",
+	)
+	require.NoError(t, err)
 }
 
 // TODO(andrei): Test that descriptor leases are released by the

--- a/pkg/sql/txn_state.go
+++ b/pkg/sql/txn_state.go
@@ -133,10 +133,12 @@ const (
 // sqlTimestamp: The timestamp to report for current_timestamp(), now() etc.
 // historicalTimestamp: If non-nil indicates that the transaction is historical
 //   and should be fixed to this timestamp.
-// priority: The transaction's priority.
+// priority: The transaction's priority. Pass roachpb.UnspecifiedUserPriority if the txn arg is
+//   not nil.
 // readOnly: The read-only character of the new txn.
 // txn: If not nil, this txn will be used instead of creating a new txn. If so,
-//      all the other arguments need to correspond to the attributes of this txn.
+//   all the other arguments need to correspond to the attributes of this txn
+//   (unless otherwise specified).
 // tranCtx: A bag of extra execution context.
 func (ts *txnState) resetForNewSQLTxn(
 	connCtx context.Context,
@@ -206,16 +208,19 @@ func (ts *txnState) resetForNewSQLTxn(
 	if txn == nil {
 		ts.mu.txn = kv.NewTxnWithSteppingEnabled(ts.Ctx, tranCtx.db, tranCtx.nodeIDOrZero)
 		ts.mu.txn.SetDebugName(opName)
+		if err := ts.setPriorityLocked(priority); err != nil {
+			panic(err)
+		}
 	} else {
+		if priority != roachpb.UnspecifiedUserPriority {
+			panic(errors.AssertionFailedf("unexpected priority when using an existing txn: %s", priority))
+		}
 		ts.mu.txn = txn
 	}
 	ts.mu.txnStart = timeutil.Now()
 	ts.mu.Unlock()
 	if historicalTimestamp != nil {
 		ts.setHistoricalTimestamp(ts.Ctx, *historicalTimestamp)
-	}
-	if err := ts.setPriority(priority); err != nil {
-		panic(err)
 	}
 	if err := ts.setReadOnlyMode(readOnly); err != nil {
 		panic(err)
@@ -299,9 +304,12 @@ func (ts *txnState) getReadTimestamp() hlc.Timestamp {
 
 func (ts *txnState) setPriority(userPriority roachpb.UserPriority) error {
 	ts.mu.Lock()
-	err := ts.mu.txn.SetUserPriority(userPriority)
-	ts.mu.Unlock()
-	if err != nil {
+	defer ts.mu.Unlock()
+	return ts.setPriorityLocked(userPriority)
+}
+
+func (ts *txnState) setPriorityLocked(userPriority roachpb.UserPriority) error {
+	if err := ts.mu.txn.SetUserPriority(userPriority); err != nil {
 		return err
 	}
 	ts.priority = userPriority

